kvserver: deflake all tests that restart test cluster servers

tbg opened the pull request:
This PR reworks how TestCluster's reusable listeners work. They are now enforced in tests that restart servers under the same listening address, since restarting without one opens the door to a race where the port is now assigned to another process.

The 23.1 backport will address #105940.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106145)
<!-- Reviewable:end -->

tbg committed:
listenerutil: export ReusableListener

We'll type assert on it in future commits.

tbg committed:
listenerutil: code movement

Future commits will change the receiver to ReusableListener.

tbg committed:
listenerutil: return ReusableListener from GetOrFail

tbg committed:
listenerutil: put registry into ReusableListener

tbg committed:
listenerutil: add ListenerRegistry.MustGet

We'll clean up the possible confusion with `GetOrFail` (which will create the listener if not present) later.

tbg committed:
listenerutil: move Reopen to ReusableListener

tbg committed:
testutils: move ListenerRegistry to leaf package

We'll want to access it from `base`, which can't import `testutils` due to
dependency cycles.

tbg committed:
listenerutil: rename GetOrFail to MustGetOrCreate

tbg committed:
listenerutil: use require.TestingT instead of *testing.T

tbg committed:
listenerutil: add a unit test

tbg committed:
listenerutil: return pointer from NewListenerRegistry

tbg committed:
testcluster: introduce ReusableListenerReg

This will replace the ReusableListeners flag and simplify all tests using it.

It's not used as of this commit, and all the tests using the soon-to-be-legacy
approach still pass.

tbg committed:
*: phase out ReusableListener bool

tbg committed:
loqrecovery: address a TODO

tbg committed:
*: adopt reusable listeners in all tests that restart

They didn't do this before, and as a result were susceptible to flakes.

They were discovered through the checks introduced in follow-up commits. It
will be impossible to hold it wrong.

erikgrinaker reviewed (approved):
Nice cleanup, thanks!

tbg commented:
TFTR!

bors r=erikgrinaker
